presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -178,14 +181,34 @@ private Expression unwrapCast(ComparisonExpression expression)
             Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
             Type targetType = typeAnalyzer.getType(session, types, expression.getRight());
 
-            if (!hasInjectiveImplicitCoercion(sourceType, targetType)) {
+            ImplicitInjectiveCoercionAbility implicitInjectiveCoercionAbility = getImplicitInjectiveCoercionAbility(sourceType, targetType);
+            if (implicitInjectiveCoercionAbility == NONE) {
                 return expression;
             }
+            if (implicitInjectiveCoercionAbility == PARTIAL) {
+                try {
+                    if (targetType.equals(DOUBLE) && sourceType.equals(BIGINT)) {
+                        Double rightDouble = (Double) right;
+                        if (rightDouble.isInfinite() || rightDouble.isNaN() || rightDouble >= 1L << 53 || rightDouble <= -1L << 53) {
+                            return expression;
```

Comment:
No need to give up here. Most of the cases could be handled:
- a comparison with NaN
- a comparison with any Double outside of Bigint range (including infinity)

Reply:
Ok, yes. I think they can be handled too : 
1. for NaN, should we just remove the check? currently I think there is no optimization for NaN - we could add some check to detect NaN and return `false` predicate. 
2. if the double value is outside the bigint range, then probably we can do a `true/false` predicate depending on the operator. are you referring to that?

Reply:
> if the double value is outside the bigint range, then probably we can do a true/false predicate depending on the operator. are you referring to that?

yes, but mind the null. There's helper methods for that `falseIfNotNull()`, `trueIfNotNull()`.

> we could add some check to detect NaN and return false predicate.

That depends on the comparison operator, too. For NOT_EQUAL and for IS_DISTINCT_FROM  it will be true. And mind the null.

Reply:
> yes, but mind the null. There's helper methods for that falseIfNotNull(), trueIfNotNull().

Ok, yes. Ideally, we can also let the optimizations done for boundary (or out of bound) values done in the rule. I think they can handle the case where double value >= 2^63 and <= -2^63. I will test that. 

>That depends on the comparison operator, too. For NOT_EQUAL and for IS_DISTINCT_FROM it will be true. And mind the null.

Yes, that's true.

Reply:
> Ok, yes. Ideally, we can also let the optimizations done for boundary (or out of bound) values done in the rule. I think they can handle the case where double value >= 2^63 and <= -2^63. I will test that.

Exactly. That I meant by "a comparison with any Double outside of Bigint range (including infinity)".

After this is done, the only uncovered areas will be those for the Double value ranging from (+/-) 2^35 until 2^63. Please add tests for all boundary values.